Feat proposal updates farcaster notifications

sshmm opened the pull request:
Add support for processing and notifying proposal updates
Introduces new functionality to:
- Fetch proposal updates via EAS GraphQL endpoints
- Process updates for followers of DAOs
- Send notifications for proposal milestone updates
- Add configuration for EAS GraphQL endpoints and schema ID
- Implement caching mechanism for fetched daos and tracking notified updates



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new development script for processing propdates.
  - Added a CLI sub-command for processing proposal updates and sending notifications.
  - Enhanced capabilities for retrieving associated proposal data from external sources.
  - Implemented new functions for handling and processing attestations related to propdates.

- **Improvements**
  - Upgraded notification handling with clearer message formatting.
  - Strengthened error handling and caching for more reliable data retrieval.
  - Expanded data structure capabilities with new interfaces for attestations and propdates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sshmm committed:
feat(propdates): Add support for processing and notifying proposal updates

Introduces new functionality to:
- Fetch proposal updates via EAS GraphQL endpoints
- Process updates for followers of DAOs
- Send notifications for proposal milestone updates
- Add configuration for EAS GraphQL endpoints and schema ID
- Implement caching mechanism for fetched daos and tracking notified updates

(cherry picked from commit 5d9ed8d20bd9d2abf3ea4c6cf40659927be3ea6d)

sshmm committed:
Filter out reply propdates and add onchain filter

dan13ram commented:
@coderabbitai summary

dan13ram commented:
@coderabbitai full review

dan13ram commented on src/services/builder/types.ts:
```diff
@@ -34,3 +38,41 @@ export interface Proposal {
+  timeCreated: number
+}
+
+export interface UpdateDao {
```
```suggestion
export interface DaoMetadata {
```

Name is somewhat misleading, its not an update to the dao but some metadata of the dao & proposal.

dan13ram commented on src/services/builder/index.ts:
```diff
@@ -27,6 +27,12 @@ export const endpoints: Record<number, string> = {
   7777777: env.BUILDER_SUBGRAPH_ZORA_URL,
 }
 
+export const attestationsEndpoints: Record<number, string> = {
```
```suggestion
export const attestationEndpoints: Record<number, string> = {
```

dan13ram commented on src/services/builder/index.ts:
```diff
@@ -37,3 +43,18 @@ export const chainEndpoints = chains.map((chain) => {
   }
 })
+
+export const attestationsChainsEndpoints = chains
```
```suggestion
export const attestationChainEndpoints = chains
```

dan13ram commented on src/services/builder/get-propdate-attestations.ts:
```diff
@@ -0,0 +1,120 @@
+      flatMap((propdates) => propdates),
+    )
+
+    console.log(propdates)
```
There are some occurences of `console.log`, which can be either removed or replaced with `logger.debug`

dan13ram commented on src/services/builder/get-propdate-attestations.ts:
```diff
@@ -0,0 +1,120 @@
+        const query = gql`
+        {
+          attestations(
+            where: {
```
Consider optimizing the number of attestations queried here. Right now, the query returns only a few, so performance isn’t an issue. However, as the response size grows—potentially reaching thousands for this `schemaId`—it could become inefficient.  

To mitigate this, we might limit results to attestations from the past week and/or the most recent 100 or 1000, sorted by `timeCreated`.

dan13ram commented on src/services/builder/get-dao-for-treasury.ts:
```diff
@@ -0,0 +1,57 @@
+    const endpoint = chainEndpoints.find(
+      (endPoint) => endPoint.chain.id === chain.id,
+    )
+    console.log(endpoint)
```
Could remove / replace `console.log` again here.

dan13ram commented on src/commands/process/propdates.ts:
```diff
@@ -0,0 +1,191 @@
+ * @returns A promise that resolves when all updates have been processed
+ * @throws Error if there's an issue fetching or processing updates
+ */
+async function handleProposalsUpdates() {
```
```suggestion
async function handleProposalUpdates() {
```

dan13ram commented on src/commands/index.ts:
```diff
@@ -110,3 +112,52 @@ export async function getUserFid() {
+ * @param propNumber - proposal number
+ * @returns A promise that resolves to the DAO object
+ */
+export async function getTreasuryDao(
```
```suggestion
export async function getDaoFromTreasury(
```

dan13ram reviewed (changes requested):
Hey @sshmm, most of my comments are minor nitpicks related to variable names and console logs.

The main suggestion is to optimize the `getAttestations` query to improve efficiency. Let me know what you think!

sshmm committed:
refactor: Remove Easscan GraphQL endpoints and update DAO type references

- Removed Easscan GraphQL endpoint configurations from .env.template and config.ts.
- Updated DAO-related type references from UpdateDao to DaoMetadata across multiple files.
- Renamed functions for clarity, changing getTreasuryDao to getDaoFromTreasury and handleProposalsUpdates to handleProposalUpdates.
- Adjusted related function calls and type definitions to reflect the changes in DAO structure.

dan13ram commented on src/services/builder/index.ts:
```diff
@@ -44,11 +47,11 @@ export const chainEndpoints = chains.map((chain) => {
-    const endpoint: string | undefined = attestationsEndpoints[chain.id]
-    const schemaId: string | undefined = env.PROPDATE_SCHEMA_ID
+    const endpoint: string | undefined = attestationEndpoints[chain.id]
+    const schemaId: string | undefined = PROPDATE_SCHEMA_ID
```
we can probably have this as `string` always ?

sshmm committed:
add cron job for processing propdates

sshmm committed:
feat(attestations): Add time filter for attestations based on creation time

- Introduced a filter to retrieve attestations created within the last 24 hours using Luxon for date handling.
- Updated GraphQL query to include a timeCreated condition for improved data relevance.

sshmm committed:
Rename PROPDATE_SCHEMA_ID to PROPDATES_SCHEMA_ID for consistency

- Updated variable name for clarity and consistency in the builder service.
- Simplified type definitions for endpoint and schemaId in attestationChainEndpoints.

dan13ram committed:
fix: ensure all attestations are notified

sshmm committed:
Merge pull request #2 from SmartInvoiceXYZ/fix/notify-all-attestations

fix: ensure all attestations are notified
